### PR TITLE
ASoC: SOF: use spin_lock_irq to protect resource

### PR DESCRIPTION
**Pull request by RanderWang**
ipc_lock is used to protect ipc message list shared by user
contexts and a workqueue. So there is no need to save interrupt
status with spin_lock_irqsave, just use spin_lock_irq.

it fixs https://github.com/thesofproject/linux/issues/422

### CODE REVIEW

**Comment by ranj063** [comment]
@RanderWang im trying to educate myself better on irq...what is the definition of a IRQ function?

**Comment by RanderWang** [comment]
> @RanderWang im trying to educate myself better on irq...what is the definition of a IRQ function?

please refer to https://geek-university.com/linux/irq-interrupt-request/. I will refine function to handler

**Comment by ranj063** [comment]
@RanderWang handler automatically clarifies what you mean here. Thanks!

**Comment by ranj063** [comment]
@RanderWang btw which interrupts doe spin_lock_irq() disable? Does it disable all interrupts?

**Comment by RanderWang** [comment]
> @RanderWang btw which interrupts doe spin_lock_irq() disable? Does it disable all interrupts?

interrupt in local cpu, not in other cpu

**Comment by RanderWang** [comment]
@keyonjie  It is strange to add these comments before spin_lock. The usage of spin_lock should be in everyone's mind. I don't see any comment to spin_lock in HDA.

**Comment by plbossart** [comment]
> @keyonjie It is strange to add these comments before spin_lock. The usage of spin_lock should be in everyone's mind. I don't see any comment to spin_lock in HDA.

@RanderWang [sarcasm] it's so clear we are at the 3rd PR to fix this... [/sarcasm] Please document what the assumptions are so that everyone can figure it out for themselves. It's not because others don't explain their code that we should follow their example, and the questions above show that this knowledge on spin_locks is not widely shared in the team.

**Comment by ranj063** [comment]
@plbossart thanks for this information: "It should clearly explain why the ipc functions are not called in IRQ handler and thus the use of spin_lock_irqsave is not needed" 

I was so confused about spin_lock_irq and spin_lock_irqsave...this makes it clear as day!

**Comment by keyonjie** [comment]
> @keyonjie It is strange to add these comments before spin_lock. The usage of spin_lock should be in everyone's mind. I don't see any comment to spin_lock in HDA.

@RanderWang I meant before the functions, not before each calling to spin_lock. similar to this one in sst-ipc.c:
```
/* locks held by caller */
void sst_ipc_tx_msg_reply_complete(struct sst_generic_ipc *ipc,
	struct ipc_message *msg)
{
	msg->complete = true;

	if (!msg->wait)
		list_add_tail(&msg->list, &ipc->empty_list);
	else
		wake_up(&msg->waitq);
}
EXPORT_SYMBOL_GPL(sst_ipc_tx_msg_reply_complete);
```

**Comment by RanderWang** [comment]
ok, I will update later

**Comment by RanderWang** [comment]
sorry, I cant connect to my work PC via VPN, so I can't  update my PR today

**Comment by RanderWang** [comment]
update PR, thanks for review

File: sound/soc/sof/ipc.c [hunk 1]
```diff
@@ -197,14 +197,18 @@ static int tx_wait_done(struct snd_sof_ipc *ipc, struct snd_sof_ipc_msg *msg,
...
 
 	/* wait for DSP IPC completion */
 	ret = wait_event_timeout(msg->waitq, msg->ipc_complete,
 				 msecs_to_jiffies(IPC_TIMEOUT_MS));
 
-	spin_lock_irqsave(&sdev->ipc_lock, flags);
+	/* interrupt is never disabled before spin lock with
+	 * ipc_lock, so spin_lock_irq is enough. And if this
+	 * condition is not met, please modify spin_lock_irq
+	 * to spin_lock_irqsave
```
Comment (plbossart) [inline sound/soc/sof/ipc.c]: Sorry @RanderWang, this does not cut it. What is needed is a clear explanation on why the conditions to use spin_lock_irq are met in the SOF IPC, and why irqsave is not needed. Your comments are unclear and raise doubts for reviewers.

Your commit message is also totally unclear. It's a circular explanation with no logical line of thought. Please reword it by explaining how the IPC is used, and based on this why spin_lock_irq is sufficient. We don't want a general lecture on spin lock, but rather an explanation of how the IPC works.

Comment (RanderWang) [reply sound/soc/sof/ipc.c]: I may focus on lecture and miss the target. Ok, I will update PR later.

**Comment by RanderWang** [comment]
update PR, thanks!

File: sound/soc/sof/ipc.c [hunk 2]
```diff
@@ -197,14 +197,17 @@ static int tx_wait_done(struct snd_sof_ipc *ipc, struct snd_sof_ipc_msg *msg,
...
-	unsigned long flags;
 	int ret;
 
 	/* wait for DSP IPC completion */
 	ret = wait_event_timeout(msg->waitq, msg->ipc_complete,
 				 msecs_to_jiffies(IPC_TIMEOUT_MS));
 
-	spin_lock_irqsave(&sdev->ipc_lock, flags);
+	/* ipc_lock is used to protect ipc message list shared by user
+	 * contexts and a workqueue. It is no need to save interrupt
```
Comment (plbossart) [inline sound/soc/sof/ipc.c]: grammar: There is no need

Comment (RanderWang) [reply sound/soc/sof/ipc.c]: yes, I will refine it

**Comment by RanderWang** [comment]
update PR, thanks!

**Pull request merged**